21.md

```diff
@@ -0,0 +1,209 @@
+- The inclusion of a `convertible currency` implies the `SERVICE` can quote and guarantee a price for given currency
+- The price returned on this response is not guaranteed by the `SERVICE` and is subject to change
+- The `code` of a `currency` will be used as an identifier for the next request
+- The `code` and other information must be according to [ISO-4217](https://en.wikipedia.org/wiki/ISO_4217) for currencies that are included in it
+- The order of the `currencies` must be interpreted by the `WALLET` as the receiving user preference for a currency
+
+### Wallet-side second request
+
+Upon seeing the `currencies` field on the response from `SERVICE`, a `WALLET` may show the user it has the option of denominating the amount in different currencies or even of sending sats to be credited as the different currency for the receiver.
+
+After gathering the input from the user, it must be encoded into the callback as query parameters as follows.
+
+The most general case, when the user is denominating the amount in `CURRENCY1`, but wants it to be converted to `CURRENCY2` on `SERVICE`.
+`CURRENCY1` and `CURRENCY2` can be the same, which would the common use case for remittances.
+```diff
+- <callback><?|&>amount=<milliSatoshi>
++ <callback><?|&>amount=<AMOUNT><CURRENCY1>&convert=<CURRENCY2>
+
+In a more particular case, when user is denominating the amount in `CURRENCY`, but does not want to convert it or the service does not support conversion.
+```diff
+- <callback><?|&>amount=<milliSatoshi>
++ <callback><?|&>amount=<AMOUNT><CURRENCY>
+```
+
+In another particular case, when the user is denominating the amount in millisatoshi, but wants it to be converted to `CURRENCY` on `SERVICE`.
+```diff
+- <callback><?|&>amount=<millisatoshi>
++ <callback><?|&>amount=<millisatoshi>&convert=<CURRENCY>
+```
+
+Note that the amount provided in all requests is always an integer number denominated in the smallets possible unit of the selected currency. The smallest unit need to be in sync with the `decimals` parameter, so the `WALLET` has all the needed information to receive input and show output properly.
+
+### Service-side second response
+
+Upon receiving a currency denominated request from `WALLET`, the `SERVICE` must return a lightning invoice with an amount matching the converted rate from currency. The rate does not need to be the same as provided on the first response.
+
+If the `WALLET` requested an actual conversion, the `SERVICE` must provide an additional field alongside the invoice informing the guaranteed `converted` amount that will be credit to the receiver upon payment. The `converted` amount, and therefore the conversion rate, must be guaranteed by the `SERVICE` for as long as the lightning invoice is not expired. The `converted` amount must be denominated in the smallest unit of the currency, just like the `amount` parameter.
+
+```typescript
+type BaseResponse = {
+  pr: string,
+  routes: [],
+}
```

Comment:
What if I request a non-supported currency? let's say `?amount=100NotSupported`

It should not be the following IMO

```json
{
	"reason": "expired",
	"status": "ERROR"
}
```

Reply:
It absolutely should not, it should probably be a more descriptive error of this format, but it can be `SERVICE` dependent, as calling the callback with unsupported currency would be a `WALLET` mistake. Does that make sense? I'll probably change my implementation to return something like ` { "status": "ERROR", "reason": "Currency not supported" }` or something.

Reply:
@jaonoctus Now I'm getting `{"reason":"unsuported currency","status":"ERROR"}` on my implementation if the currency is not supported. It's an improvement